### ISSUE TITLE
Greedy break in ZERO

### ISSUE THREAD

**Issue by Bostoncake** [FarinaMatteo/zero#2]
Hi! Thank you so much for releasing the code for the reproducibility of your published results! I have a question about the greedy break formula inside ZERO. In https://github.com/FarinaMatteo/zero/blob/1bbb51a68ad9a5569979bc1a38da46e238c9628f/ttas/zero.py#L83, `l[k:]` is used for breaking the tie. However, `l` here is arranged in the original augmented image list order, so there could be filtered views in `l[k:]`. Maybe you should sort `l` in the order of their corresponding entropy, and then use `l[k:]`. Is this modification correct for your ZERO implementation?

Thanks!

**Comment by FarinaMatteo** [comment]
Hello @Bostoncake, good catch, and thanks for pointing this out! 

I have probably introduced this while re-factoring the code back in May. Your interpretation is indeed correct, the logits should be sorted in order of ascending entropy before `greedy_break` (most to least confident). This also holds for `ttas/zero_rlcf.py`. 

I plan to update the repo within a couple of weeks since I will also update the article on arXiv. I will make sure to fix this by then.  

All the best

**Comment by FarinaMatteo** [comment]
> Hello @Bostoncake, good catch, and thanks for pointing this out!
> 
> I have probably introduced this while re-factoring the code back in May. Your interpretation is indeed correct, the logits should be sorted in order of ascending entropy before `greedy_break` (most to least confident). This also holds for `ttas/zero_rlcf.py`.
> 
> I plan to update the repo within a couple of weeks since I will also update the article on arXiv. I will make sure to fix this by then.
> 
> All the best

Hello again @Bostoncake.
The repo as well as the article on arxiv have now been updated, and the bug has been fixed. Please let me know if everything's fine, so I can safely close this issue. Thank you once more for pointing this out. Wish you a good day!

**Comment by Bostoncake** [comment]
> > Hello @Bostoncake, good catch, and thanks for pointing this out!
> > I have probably introduced this while re-factoring the code back in May. Your interpretation is indeed correct, the logits should be sorted in order of ascending entropy before `greedy_break` (most to least confident). This also holds for `ttas/zero_rlcf.py`.
> > I plan to update the repo within a couple of weeks since I will also update the article on arXiv. I will make sure to fix this by then.
> > All the best
> 
> Hello again @Bostoncake. The repo as well as the article on arxiv have now been updated, and the bug has been fixed. Please let me know if everything's fine, so I can safely close this issue. Thank you once more for pointing this out. Wish you a good day!

Everything is fine now for me. You can close this issue if you like @FarinaMatteo. Thank you for your inspiring work and quick response!
